File: packages/py/polywrap-wasm/polywrap_wasm/imports.py
from unsync import unsync, Unfuture
from wasmtime import (
    Store,
    Linker,
    Memory,
    Module,
    MemoryType,
    Limits,
    ValType,
    FuncType,
    Instance,
)

from polywrap_core import Invoker, InvokerOptions, Uri, InvokeResult

from .buffer import write_string, write_bytes, read_string, read_bytes
from .types.state import State
from .errors import WasmAbortError


@unsync
async def unsync_invoke(invoker: Invoker, options: InvokerOptions) -> InvokeResult:
    return await invoker.invoke(options)


def create_instance(
    store: Store,
    module: Module,
    state: State,
    invoker: Invoker,
) -> Instance:
    linker = Linker(store.engine)

    """
    TODO: Re-check this based on issue https://github.com/polywrap/toolchain/issues/561
    This probably means that memory creation should be moved to its own function 
    """
    mem = Memory(store, MemoryType(Limits(1, None)))

    wrap_abort_type = FuncType(
        [
            ValType.i32(),
            ValType.i32(),
            ValType.i32(),
            ValType.i32(),
            ValType.i32(),
            ValType.i32(),
        ],
        [],
    )

    def wrap_abort(
        msg_offset: int,
        msg_len: int,
        file_offset: int,
        file_len: int,
        line: int,
        column: int,
    ) -> None:
        msg = read_string(mem.data_ptr(store), mem.data_len(store), msg_offset, msg_len)
        file = read_string(
            mem.data_ptr(store), mem.data_len(store), file_offset, file_len
        )
        raise WasmAbortError(
            f"__wrap_abort: {msg}\nFile: {file}\nLocation: [{line},{column}]"
        )

    wrap_invoke_args_type = FuncType([ValType.i32(), ValType.i32()], [])

    def wrap_invoke_args(method_ptr: int, args_ptr: int) -> None:
        if not state.method:
            raise WasmAbortError("__wrap_invoke_args: method is not set")
        else:
            write_string(
                mem.data_ptr(store), mem.data_len(store), state.method, method_ptr
            )

        if not state.args:
            raise WasmAbortError("__wrap_invoke_args: args is not set")
        else:
            write_bytes(
                mem.data_ptr(store),
                mem.data_len(store),
                bytearray(state.args),
                args_ptr,
            )

    wrap_invoke_result_type = FuncType([ValType.i32(), ValType.i32()], [])

    def wrap_invoke_result(offset: int, length: int) -> None:
        result = read_bytes(mem.data_ptr(store), mem.data_len(store), offset, length)
        state.invoke["result"] = result

    wrap_invoke_error_type = FuncType([ValType.i32(), ValType.i32()], [])

    def wrap_invoke_error(offset: int, length: int):
        error = read_string(mem.data_ptr(store), mem.data_len(store), offset, length)
        state.invoke["error"] = error

    wrap_subinvoke_type = FuncType(
        [
            ValType.i32(),
            ValType.i32(),
            ValType.i32(),
            ValType.i32(),
            ValType.i32(),
            ValType.i32(),
        ],
        [ValType.i32()],
    )

    def wrap_subinvoke(
        uri_ptr: int,
        uri_len: int,
        method_ptr: int,
        method_len: int,
        args_ptr: int,
        args_len: int,
    ) -> bool:
        uri = read_string(mem.data_ptr(store), mem.data_len(store), uri_ptr, uri_len)
        method = read_string(
            mem.data_ptr(store), mem.data_len(store), method_ptr, method_len
        )
        args = read_bytes(mem.data_ptr(store), mem.data_len(store), args_ptr, args_len)

        unfuture_result: Unfuture[InvokeResult] = unsync_invoke(
            invoker,
            InvokerOptions(uri=Uri(uri), method=method, args=args, encode_result=True),
        )
        result = unfuture_result.result()

        if result.result:
            state.subinvoke["result"] = result.result
            return True
        elif result.error:
            state.subinvoke["error"] = "".join(str(x) for x in result.error.args)
            return False
        else:
            raise ValueError("subinvocation failed!")

    wrap_subinvoke_result_len_type = FuncType([], [ValType.i32()])

    def wrap_subinvoke_result_len() -> int:
        if not state.subinvoke["result"]:
            raise WasmAbortError(
                "__wrap_subinvoke_result_len: subinvoke.result is not set"
            )
        return len(state.subinvoke["result"])

    wrap_subinvoke_result_type = FuncType([ValType.i32()], [])

    def wrap_subinvoke_result(ptr: int) -> None:
        if not state.subinvoke["result"]:
            raise WasmAbortError("__wrap_subinvoke_result: subinvoke.result is not set")
        write_bytes(
            mem.data_ptr(store), mem.data_len(store), state.subinvoke["result"], ptr
        )

    wrap_subinvoke_error_len_type = FuncType([], [ValType.i32()])

    def wrap_subinvoke_error_len() -> int:
        if not state.subinvoke["error"]:
            raise WasmAbortError(
                "__wrap_subinvoke_error_len: subinvoke.error is not set"
            )
        return len(state.subinvoke["error"])

    wrap_subinvoke_error_type = FuncType([ValType.i32()], [])

    def wrap_subinvoke_error(ptr: int) -> None:
        if not state.subinvoke["error"]:
            raise WasmAbortError("__wrap_subinvoke_error: subinvoke.error is not set")
        write_string(
            mem.data_ptr(store), mem.data_len(store), state.subinvoke["error"], ptr
        )

    # TODO: use generics or any on wasmtime codebase to fix typings
    linker.define_func("wrap", "__wrap_abort", wrap_abort_type, wrap_abort)  # type: ignore partially unknown
    linker.define_func("wrap", "__wrap_invoke_args", wrap_invoke_args_type, wrap_invoke_args)  # type: ignore partially unknown
    linker.define_func("wrap", "__wrap_invoke_result", wrap_invoke_result_type, wrap_invoke_result)  # type: ignore partially unknown
    linker.define_func("wrap", "__wrap_invoke_error", wrap_invoke_error_type, wrap_invoke_error)  # type: ignore partially unknown
    linker.define_func("wrap", "__wrap_subinvoke", wrap_subinvoke_type, wrap_subinvoke)  # type: ignore partially unknown
    linker.define_func("wrap", "__wrap_subinvoke_result_len", wrap_subinvoke_result_len_type, wrap_subinvoke_result_len)  # type: ignore partially unknown
    linker.define_func("wrap", "__wrap_subinvoke_result", wrap_subinvoke_result_type, wrap_subinvoke_result)  # type: ignore partially unknown
    linker.define_func("wrap", "__wrap_subinvoke_error_len", wrap_subinvoke_error_len_type, wrap_subinvoke_error_len)  # type: ignore partially unknown
    linker.define_func("wrap", "__wrap_subinvoke_error", wrap_subinvoke_error_type, wrap_subinvoke_error)  # type: ignore partially unknown
    linker.define("env", "memory", mem)
    return linker.instantiate(store, module)

File: packages/js/wasm/src/imports.ts
/* eslint-disable @typescript-eslint/naming-convention */

import { u32, WrapImports } from "./types";
import { readBytes, readString, writeBytes, writeString } from "./buffer";
import { State } from "./WasmWrapper";

import { msgpackEncode } from "@polywrap/msgpack-js";
import { CoreClient, Uri, ErrorSource } from "@polywrap/core-js";

export const createImports = (config: {
  client: CoreClient;
  memory: WebAssembly.Memory;
  state: State;
  abortWithInvokeAborted: (message: string, source: ErrorSource) => never;
  abortWithInternalError: (message: string) => never;
}): WrapImports => {
  const {
    memory,
    state,
    client,
    abortWithInvokeAborted,
    abortWithInternalError,
  } = config;

  return {
    wrap: {
      __wrap_subinvoke: async (
        uriPtr: u32,
        uriLen: u32,
        methodPtr: u32,
        methodLen: u32,
        argsPtr: u32,
        argsLen: u32
      ): Promise<boolean> => {
        // Reset our state
        state.subinvoke.result = undefined;
        state.subinvoke.error = undefined;

        const uri = readString(memory.buffer, uriPtr, uriLen);
        const method = readString(memory.buffer, methodPtr, methodLen);
        const args = readBytes(memory.buffer, argsPtr, argsLen);

        const result = await client.invoke<Uint8Array>({
          uri: Uri.from(uri),
          method: method,
          args: new Uint8Array(args),
          encodeResult: true,
        });

        if (result.ok) {
          state.subinvoke.result = result.value;
        } else {
          state.subinvoke.error = `${result.error?.name}: ${result.error?.message}`;
        }

        return result.ok;
      },
      // Give WASM the size of the result
      __wrap_subinvoke_result_len: (): u32 => {
        if (!state.subinvoke.result) {
          abortWithInternalError(
            "__wrap_subinvoke_result_len: subinvoke.result is not set"
          );
          return 0;
        }
        return state.subinvoke.result.byteLength;
      },
      // Copy the subinvoke result into WASM
      __wrap_subinvoke_result: (ptr: u32): void => {
        if (!state.subinvoke.result) {
          abortWithInternalError(
            "__wrap_subinvoke_result: subinvoke.result is not set"
          );
          return;
        }
        writeBytes(state.subinvoke.result, memory.buffer, ptr);
      },
      // Give WASM the size of the error
      __wrap_subinvoke_error_len: (): u32 => {
        if (!state.subinvoke.error) {
          abortWithInternalError(
            "__wrap_subinvoke_error_len: subinvoke.error is not set"
          );
          return 0;
        }
        return state.subinvoke.error.length;
      },
      // Copy the subinvoke error into WASM
      __wrap_subinvoke_error: (ptr: u32): void => {
        if (!state.subinvoke.error) {
          abortWithInternalError(
            "__wrap_subinvoke_error: subinvoke.error is not set"
          );
          return;
        }
        writeString(state.subinvoke.error, memory.buffer, ptr);
      },
      __wrap_subinvokeImplementation: async (
        interfaceUriPtr: u32,
        interfaceUriLen: u32,
        implUriPtr: u32,
        implUriLen: u32,
        methodPtr: u32,
        methodLen: u32,
        argsPtr: u32,
        argsLen: u32
      ): Promise<boolean> => {
        state.subinvokeImplementation.result = undefined;
        state.subinvokeImplementation.error = undefined;

        const implUri = readString(memory.buffer, implUriPtr, implUriLen);
        const method = readString(memory.buffer, methodPtr, methodLen);
        const args = readBytes(memory.buffer, argsPtr, argsLen);

        state.subinvokeImplementation.args = [implUri, method, args];

        const result = await client.invoke<Uint8Array>({
          uri: Uri.from(implUri),
          method: method,
          args: new Uint8Array(args),
          encodeResult: true,
        });

        if (result.ok) {
          state.subinvokeImplementation.result = result.value;
        } else {
          state.subinvokeImplementation.error = `${result.error?.name}: ${result.error?.message}`;
        }

        return result.ok;
      },
      __wrap_subinvokeImplementation_result_len: (): u32 => {
        if (!state.subinvokeImplementation.result) {
          abortWithInternalError(
            "__wrap_subinvokeImplementation_result_len: subinvokeImplementation.result is not set"
          );
          return 0;
        }
        return state.subinvokeImplementation.result.byteLength;
      },
      __wrap_subinvokeImplementation_result: (ptr: u32): void => {
        if (!state.subinvokeImplementation.result) {
          abortWithInternalError(
            "__wrap_subinvokeImplementation_result: subinvokeImplementation.result is not set"
          );
          return;
        }
        writeBytes(state.subinvokeImplementation.result, memory.buffer, ptr);
      },
      __wrap_subinvokeImplementation_error_len: (): u32 => {
        if (!state.subinvokeImplementation.error) {
          abortWithInternalError(
            "__wrap_subinvokeImplementation_error_len: subinvokeImplementation.error is not set"
          );
          return 0;
        }
        return state.subinvokeImplementation.error.length;
      },
      __wrap_subinvokeImplementation_error: (ptr: u32): void => {
        if (!state.subinvokeImplementation.error) {
          abortWithInternalError(
            "__wrap_subinvokeImplementation_error: subinvokeImplementation.error is not set"
          );
          return;
        }
        writeString(state.subinvokeImplementation.error, memory.buffer, ptr);
      },
      // Copy the invocation's method & args into WASM
      __wrap_invoke_args: (methodPtr: u32, argsPtr: u32): void => {
        if (!state.method) {
          abortWithInternalError("__wrap_invoke_args: method is not set");
          return;
        }
        if (!state.args) {
          abortWithInternalError("__wrap_invoke_args: args is not set");
          return;
        }
        writeString(state.method, memory.buffer, methodPtr);
        writeBytes(state.args, memory.buffer, argsPtr);
      },
      // Store the invocation's result
      __wrap_invoke_result: (ptr: u32, len: u32): void => {
        state.invoke.result = new Uint8Array(
          readBytes(memory.buffer, ptr, len)
        );
      },
      // Store the invocation's error
      __wrap_invoke_error: (ptr: u32, len: u32): void => {
        state.invoke.error = readString(memory.buffer, ptr, len);
      },
      __wrap_getImplementations: async (
        uriPtr: u32,
        uriLen: u32
      ): Promise<boolean> => {
        const uri = readString(memory.buffer, uriPtr, uriLen);
        const result = await client.getImplementations(Uri.from(uri), {});
        if (!result.ok) {
          abortWithInternalError(result.error?.message as string);
          return false;
        }
        const implementations = result.value.map((i) => i.uri);
        state.getImplementationsResult = msgpackEncode(implementations);
        return implementations.length > 0;
      },
      __wrap_getImplementations_result_len: (): u32 => {
        if (!state.getImplementationsResult) {
          abortWithInternalError(
            "__wrap_getImplementations_result_len: result is not set"
          );
          return 0;
        }
        return state.getImplementationsResult.byteLength;
      },
      __wrap_getImplementations_result: (ptr: u32): void => {
        if (!state.getImplementationsResult) {
          abortWithInternalError(
            "__wrap_getImplementations_result: result is not set"
          );
          return;
        }
        writeBytes(state.getImplementationsResult, memory.buffer, ptr);
      },
      __wrap_load_env: (ptr: u32): void => {
        writeBytes(state.env, memory.buffer, ptr);
      },
      __wrap_abort: (
        msgPtr: u32,
        msgLen: u32,
        filePtr: u32,
        fileLen: u32,
        line: u32,
        column: u32
      ): void => {
        const msg = readString(memory.buffer, msgPtr, msgLen);
        const file = readString(memory.buffer, filePtr, fileLen);

        abortWithInvokeAborted(`__wrap_abort: ${msg}`, {
          file,
          row: line,
          col: column,
        });
      },
      __wrap_debug_log: (ptr: u32, len: u32): void => {
        const msg = readString(memory.buffer, ptr, len);
        console.debug(`__wrap_debug_log: ${msg}`);
      },
    },
    env: {
      memory,
    },
  };
};
